### ISSUE TITLE
Remove RSS service plan ledgers

### ISSUE THREAD

**Issue by andrewjstone** [oxidecomputer/omicron#7174]
When running RSS a service plan gets created to dictate where specific services run on the rack and what their configurations are in terms of datasets, IP addresses, etc.. After the service plan is created it is persisted to a ledger on the RSS node prior to deploying service zones by communicating with other bootstrap agents. 

The goal of the service plan ledger is to allow RSS to pick up where it left off if it crashes. However, currently, and for the foreseeable future we cannot actually pick up where we left off. Therefore, the service plan ledger serves no actual use. It is saved during RSS, and never read in the normal path once RSS completes. 

Normally, the existence of a small JSON file persisted on M.2 drives that never gets read is not a big deal. However, we are now up to our 5th version of the file, even though there is no need for backwards compatibility, because you can't actually rerun RSS. Anytime the structure of the service plan changes, a developer has to decide what to do, and the default is to go ahead and add another version with a backwards compatibility check. This results in unnecessary growth of a code base and slower development.

I propose removing this ledger altogether and only maintaining the in memory copy of the service plan during RSS.

We can almost certainly do the same with the sled plan for the same reasons. Notably, there is only one version of the sled plan ledger. Perhaps this is because it just never had a need to change after our initial release.

@jgallagher @sunshowers and the rest of us discussed this in update huddle, but it deserves consideration from @smklein and @davepacheco.

**Comment by davepacheco** [comment]
Sounds fine to me -- except that at one point I think the existence of the file was how RSS would know that it had run before and could not proceed safely.  As long as it has some other way to know that, this sounds good.

**Comment by smklein** [comment]
Yeah I think the most important bit is the establishment of "did RSS run to completion or not". If the answer is no, we can attempt to clear everything out and restart. If the answer is yes, then we need to respect it, and not try to wrangle control away from the running Nexus(es).

However, this should exist in the form of a "marker file": https://github.com/oxidecomputer/omicron/blob/fd50d18515587d72fd6ce4d78a35c2156a2e5900/sled-agent/src/rack_setup/service.rs#L14-L21

Anyway - if we have a nice way to "reset the world" when RSS runs, after a reboot -- if it has not fully completed before -- I'm on-board with destroying these intermediate files.

**Comment by jgallagher** [comment]
> Anyway - if we have a nice way to "reset the world" when RSS runs, after a reboot -- if it has not fully completed before -- I'm on-board with destroying these intermediate files.

We don't have a nice way, really, although we do need one. There are a whole pile of related issues here, including at least #3818, #3820, #4205, #4399, #4578, #4587, #6906. We have a not-particularly-nice way (manually clean-slate the rack), which we have done during real setups and do all the time in dev environments. I think building the nice way shouldn't block this issue, because of the point Andrew noted (which is confirmed by all these issues) - even though RSS writes out this plan so it can try to resume if it reboots partway through, that doesn't actually work and still needs a clean-slate to recover.

> Sounds fine to me -- except that at one point I think the existence of the file was how RSS would know that it had run before and could not proceed safely. As long as it has some other way to know that, this sounds good.

We probably should do this as part of this, I think? The first thing RSS does if you try to run it is check for the "RSS completed" marker file and then [hand off to Nexus](https://github.com/oxidecomputer/omicron/blob/fd50d18515587d72fd6ce4d78a35c2156a2e5900/sled-agent/src/rack_setup/service.rs#L1207-L1248). (I'm not sure if that really works? We'd need to have finished RSS but not handed off to Nexus, which I'm not sure we've tested any time recently.)

The next thing it does is either [load or create the sled plan](https://github.com/oxidecomputer/omicron/blob/fd50d18515587d72fd6ce4d78a35c2156a2e5900/sled-agent/src/rack_setup/service.rs#L1263-L1275) - if we're removing the load, can we replace it with "write an RSS-has-started marked and fail if it already exists"?

**Comment by andrewjstone** [comment]
> We probably should do this as part of this, I think? The first thing RSS does if you try to run it is check for the "RSS completed" marker file and then [hand off to Nexus](https://github.com/oxidecomputer/omicron/blob/fd50d18515587d72fd6ce4d78a35c2156a2e5900/sled-agent/src/rack_setup/service.rs#L1207-L1248). (I'm not sure if that really works? We'd need to have finished RSS but not handed off to Nexus, which I'm not sure we've tested any time recently.)

Interestingly, this path is entirely correct. We write the marker **before** we handoff to nexus. If RSS dies in between those two points the handoff will be tried again and should succeed. This prevents us from having to re-run RSS unnecessarily. If the handoff did succeed and we rebooted, the handoff is a no-op. We actually need the persisted plans if we want to keep this behavior - allowing handoff after RSS has run on a very poorly timed crash or reboot. If we keep this I'd recommend just removing the versioning from the plans but still persisting them.

The simpler alternative is to move the writing of the marker to after handoff completes and ignore this failure mode. We would write an RSS has started marker as John suggested at the start of RSS so we can bail early with an error in case RSS is run again.

All of this discussion of markers though brings me to a separate question. If the cable is swapped between scrimlets we won't see the markers at all on a re-run of RSS. After a while of running, scrimlets themselves can get swapped and the marker file not present on any sled. Should we actually have another mechanism to prevent RSS from running long after RSS was run initially and the system is up and running?

In summary, my suggestion is the following:

1. Remove the plan ledgers
2. Add an RSS started marker so we can tell if it has failed on a disconnection from wicket and require a clean-slate.
3. Move the RSS completed marker to after handoff and ignore the possibility of a crash after this marker is written and handoff completes
4. Open an issue to fix the fact that the ledger is only written to one scrimlet. To fix this I suggest writing to the bootstore a new key that says handoff has completed and then gating RSS runs on that rather than or in addition to a ledger.

**Comment by andrewjstone** [comment]
Resolved by #7185
